File: client/Core/Models/DeviceHardwareInfo.cs
using System.Text.Json;

namespace client.Core.Models;

public class DeviceHardwareInfo
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    public string MacAddress { get; set; } = string.Empty;
    public string Hostname { get; set; } = string.Empty;
    public string OsName { get; set; } = string.Empty;
    public string OsVersion { get; set; } = string.Empty;
    public string CpuModel { get; set; } = string.Empty;
    public int CpuCores { get; set; }
    public long RamTotalMb { get; set; }

    public string GpuModel { get; set; } = string.Empty;
    public long GpuVramMb { get; set; }
    public string StorageDevices { get; set; } = "[]";
    public DateTime CollectedAt { get; set; } = DateTime.UtcNow;
    public bool IsSynced { get; set; }
    public string? SyncedAt { get; set; }
    public string CreatedAt { get; set; } = string.Empty;
}

public class InstalledApplication
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    public string AppName { get; set; } = string.Empty;
    /// <summary>Executable binary name (e.g., \"code\" for Visual Studio Code)</summary>
    public string BinaryName { get; set; } = string.Empty;
    public string AppVersion { get; set; } = string.Empty;
    public string Publisher { get; set; } = string.Empty;
    public string InstallPath { get; set; } = string.Empty;
    public DateTime? InstallDate { get; set; }
    public string UninstallString { get; set; } = string.Empty;
    /// <summary>"installed" | "uninstalled" | "seen"</summary>
    public string ChangeType { get; set; } = "seen";
    /// <summary>True if this app is a web browser (detected from .desktop Categories/MimeType)</summary>
    public bool IsBrowser { get; set; }
    /// <summary>
    /// Stable desktop entry identity on Linux: the .desktop filename without extension
    /// (e.g. "firefox_firefox", "code"). Used by SoftwareIdentityResolver to dedup across
    /// package managers and detect the same app discovered via multiple sources.
    /// Empty on Windows/macOS.
    /// </summary>
    public string DesktopId { get; set; } = string.Empty;
    /// <summary>
    /// Desktop entry Categories on Linux (e.g. "WebBrowser;Network;GTK;") or macOS bundle
    /// identifier (e.g. "org.mozilla.firefox"). Used for metadata-driven classification
    /// instead of hardcoded name lists. Empty on Windows.
    /// </summary>
    public string Categories { get; set; } = string.Empty;
    public DateTime DetectedAt { get; set; } = DateTime.UtcNow;
    public bool IsSynced { get; set; }
    public string? SyncedAt { get; set; }
    public string CreatedAt { get; set; } = string.Empty;
}

public class NetworkInfo
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    public string PublicIp { get; set; } = string.Empty;
    public string PrivateIp { get; set; } = string.Empty;
    public string MacAddress { get; set; } = string.Empty;
    public string NetworkInterfaceName { get; set; } = string.Empty;
    public DateTime CollectedAt { get; set; } = DateTime.UtcNow;
    /// <summary>When this IP combination was first observed (row created).</summary>
    public DateTime? FirstSeenAt { get; set; }
    /// <summary>Last time the same IP combination was still seen as the active one.</summary>
    public DateTime? LastSeenAt { get; set; }
    /// <summary>True while this IP combination is the device's active one (flipped to 0 on IP change).</summary>
    public bool IsCurrent { get; set; } = true;
    public bool IsSynced { get; set; }
    public string? SyncedAt { get; set; }
    public string CreatedAt { get; set; } = string.Empty;
}

public class StorageDevice
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    /// <summary>FK → device_hardware_info.id</summary>
    public string DeviceHardwareId { get; set; } = string.Empty;
    public string DeviceType { get; set; } = string.Empty;
    public string Model { get; set; } = string.Empty;
    public long CapacityMb { get; set; }
    public bool IsSynced { get; set; }
    public string? SyncedAt { get; set; }
    public string CreatedAt { get; set; } = string.Empty;
}

public class InstalledPackage
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    public string PackageName { get; set; } = string.Empty;
    public string Version { get; set; } = string.Empty;
    /// <summary>"runtime" | "tool" | "library" | "system"</summary>
    public string Category { get; set; } = "tool";
    /// <summary>"npm" | "pip" | "apt" | "brew" | "choco" | "winget" | "scoop" | "cargo" | "snap" | "flatpak"</summary>
    public string SourceManager { get; set; } = string.Empty;
    public string InstallPath { get; set; } = string.Empty;
    public string Publisher { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public DateTime DetectedAt { get; set; } = DateTime.UtcNow;
    public bool IsSynced { get; set; }
    public string? SyncedAt { get; set; }
    public string CreatedAt { get; set; } = string.Empty;
}

public class SessionEvent
{
    public string Id { get; set; } = Guid.NewGuid().ToString("N");
    /// <summary>"login" | "logout" | "lock" | "unlock"</summary>
    public string EventType { get; set; } = string.Empty;
    public string OsUsername { get; set; } = string.Empty;
    public DateTime EventAt { get; set; } = DateTime.UtcNow;
    public bool IsSynced { get; set; }
    public string? SyncedAt { get; set; }
    public string CreatedAt { get; set; } = string.Empty;
}

File: server/internal/models/device_hardware_info.go
package models

import "time"

type DeviceHardwareInfo struct {
	ID             string    `json:"id" db:"id"`
	EmployeeID     string    `json:"employeeId" db:"employee_id"`
	DeviceID       string    `json:"deviceId" db:"device_id"`
	MacAddress     string    `json:"macAddress" db:"mac_address"`
	Hostname       string    `json:"hostname" db:"hostname"`
	OsName         string    `json:"osName" db:"os_name"`
	OsVersion      string    `json:"osVersion" db:"os_version"`
	CpuModel       string    `json:"cpuModel" db:"cpu_model"`
	CpuCores       int       `json:"cpuCores" db:"cpu_cores"`
	RamTotalMB     int64     `json:"ramTotalMb" db:"ram_total_mb"`
	StorageDevices string    `json:"storageDevices" db:"storage_devices"`
	GpuModel       string    `json:"gpuModel" db:"gpu_model"`
	GpuVramMB      int64     `json:"gpuVramMb" db:"gpu_vram_mb"`
	CollectedAt    time.Time `json:"collectedAt" db:"collected_at"`
	SyncedAt       *time.Time `json:"syncedAt,omitempty" db:"synced_at"`
	CreatedAt      time.Time `json:"createdAt" db:"created_at"`
}

type InstalledApplication struct {
	ID              string     `json:"id" db:"id"`
	EmployeeID      string     `json:"employeeId" db:"employee_id"`
	AppName         string     `json:"appName" db:"app_name"`
	AppVersion      string     `json:"appVersion" db:"app_version"`
	Publisher       string     `json:"publisher" db:"publisher"`
	InstallPath     string     `json:"installPath" db:"install_path"`
	InstallDate     *time.Time `json:"installDate,omitempty" db:"install_date"`
	UninstallString string     `json:"uninstallString" db:"uninstall_string"`
	ChangeType      string     `json:"changeType" db:"change_type"`
	DetectedAt      time.Time  `json:"detectedAt" db:"detected_at"`
	SyncedAt        *time.Time `json:"syncedAt,omitempty" db:"synced_at"`
	CreatedAt       time.Time  `json:"createdAt" db:"created_at"`
	BinaryName      string     `json:"binaryName,omitempty" db:"binary_name"`
	IsBrowser       bool       `json:"isBrowser" db:"is_browser"`
	DesktopID       string     `json:"desktopId,omitempty" db:"desktop_id"`
	Categories      string     `json:"categories,omitempty" db:"categories"`
	AppFingerprint  string     `json:"-" db:"app_fingerprint"`
}

type NetworkInfo struct {
	ID                   string    `json:"id" db:"id"`
	EmployeeID           string    `json:"employeeId" db:"employee_id"`
	PublicIP             string    `json:"publicIp" db:"public_ip"`
	PrivateIP            string    `json:"privateIp" db:"private_ip"`
	MacAddress           string    `json:"macAddress" db:"mac_address"`
	NetworkInterfaceName string    `json:"networkInterfaceName" db:"network_interface_name"`
	CollectedAt          time.Time `json:"collectedAt" db:"collected_at"`
	SyncedAt             *time.Time `json:"syncedAt,omitempty" db:"synced_at"`
	CreatedAt            time.Time `json:"createdAt" db:"created_at"`
}

type InstalledPackage struct {
	ID            string     `json:"id" db:"id"`
	EmployeeID    string     `json:"employeeId" db:"employee_id"`
	PackageName   string     `json:"packageName" db:"package_name"`
	Version       string     `json:"version" db:"version"`
	Category      string     `json:"category" db:"category"`
	SourceManager string     `json:"sourceManager" db:"source_manager"`
	InstallPath   string     `json:"installPath" db:"install_path"`
	Publisher     string     `json:"publisher" db:"publisher"`
	Description   string     `json:"description" db:"description"`
	DetectedAt    time.Time  `json:"detectedAt" db:"detected_at"`
	SyncedAt      *time.Time `json:"syncedAt,omitempty" db:"synced_at"`
	CreatedAt     time.Time  `json:"createdAt" db:"created_at"`
	PackageFingerprint string `json:"-" db:"package_fingerprint"`
}

type SessionEvent struct {
	ID         string    `json:"id" db:"id"`
	EmployeeID string    `json:"employeeId" db:"employee_id"`
	EventType  string    `json:"eventType" db:"event_type"`
	OsUsername string    `json:"osUsername" db:"os_username"`
	EventAt    time.Time `json:"eventAt" db:"event_at"`
	SyncedAt   *time.Time `json:"syncedAt,omitempty" db:"synced_at"`
	CreatedAt  time.Time `json:"createdAt" db:"created_at"`
}

